### ISSUE TITLE
Removal of the root

### ISSUE THREAD

**Issue by espadrine** [json-patch/json-patch2#17]
The patch `{ "op": "remove", "path": "" }` is currently undefined.

I believe it should have the same result as `{ "op": "add", "path": "", "value": null }`. As a side-result, it would also validate the definition of `{ "op": "replace", "path": "", "value": null }`, which states:

> This operation is functionally identical to a "remove" operation for
>   a value, followed immediately by an "add" operation at the same
>   location with the replacement value.

In any case, it would benefit from being specified. [Some implementations](https://tonicdev.com/577dbcec8833121300d6e5d3/57bc7b45c8db44140088cdfa/branches/master) currently throw, and that is unfortunate, as it breaks the specification as quoted above..

**Comment by d-frey** [comment]
I disagree with the solution. I think `{ "op": "remove", "path": "" }` should be an error. If you want to replace the root element, you can already use `{ "op": "add", "path": "", "value": null }` today. Even that is strange.

It could be reasonable to allow the `replace` to work with an empty path to replace the element completely, and, compatibility aside, disallow `add` to do the same. This would provide the most intuitive semantics. `add` and `remove` require a non-empty path, `replace` is able to replace the root.

But since we most likely want to be compatible, I think `add` should stay as it is, `replace` should explicitly allow to replace the root and `remove` on the root should be an error.

**Comment by espadrine** [comment]
What matters most is that the undefined behavior gets specified.

That being said, if we go with your solution, we lose the functional equivalence of `replace` to `remove` + `add`, which will be a bit counterintuitive.
